text-decoration-line: Refactor TextDecorationLine into strong type

vitorroriz opened the pull request:
#### b0dc25a5f71ee46f06838c3694d65c9489f15be3
<pre>
text-decoration-line: Refactor TextDecorationLine into strong type
<a href="https://bugs.webkit.org/show_bug.cgi?id=297756">https://bugs.webkit.org/show_bug.cgi?id=297756</a>
<a href="https://rdar.apple.com/158897260">rdar://158897260</a>

Reviewed by Tim Nguyen and Simon Fraser.

text-decoration-line used to be represented on Style as a OptionSet&lt;TextDecorationLine&gt;
and TextDecorationLine used to be an enum class.

We are refactoring TextDecorationLine into its own type aligned with other strong type changes.
The new type is now defined in StyleTextDecorationLine and it enforces the grammar:

text-decoration-line = none | [ underline || overline || line-through || blink ] | spelling-error | grammar-error

The new underlying representation for it is:

Variant&lt;CSS::Keyword::None, CSS::Keyword::SpellingError, CSS::Keyword::GrammarError, OptionSet&lt;TextDecorationLineFlags&gt;&gt; m_value;

This means that TextDecorationLine flags represents now only [ underline || overline || line-through || blink ]

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::LineDecorationStyle::LineDecorationStyle):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onStyleChange):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::hasPlainText const):
(WebCore::AccessibilityRenderObject::hasUnderline const):
* Source/WebCore/accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::textAttributes const):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::attributeStringSetStyle):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::fontAttributesAtSelectionStart):
* Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
(WebCore::updateAttributes):
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
(WebCore::Layout::logicalBottomForTextDecorationContent):
(WebCore::Layout::InlineDisplayContentBuilder::collectInkOverflowForTextDecorations):
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForHighlights):
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::computeStyleForPseudoElementStyle):
(WebCore::computeStylesForTextDecorations):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter::paintForegroundAndDecorations):
(WebCore::computedTextDecorationType):
(WebCore::isDecoratingBoxForBackground):
(WebCore::TextBoxPainter::paintBackgroundDecorations):
(WebCore::TextBoxPainter::paintForegroundDecorations):
(WebCore::TextBoxPainter::paintPlatformDocumentMarkers):
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::TextDecorationPainter::paintBackgroundDecorations):
(WebCore::collectStylesForRenderer):
(WebCore::TextDecorationPainter::stylesForRenderer):
(WebCore::TextDecorationPainter::textDecorationsInEffectForStyle):
* Source/WebCore/rendering/TextDecorationPainter.h:
(WebCore::TextDecorationPainter::stylesForRenderer):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
(WebCore::RenderStyle::NonInheritedFlags::copyNonInheritedFrom):
(WebCore::RenderStyle::changeAffectsVisualOverflow const):
(WebCore::miscDataChangeRequiresRepaint):
(WebCore::RenderStyle::changeRequiresRepaint const):
(WebCore::RenderStyle::changeRequiresRepaintIfText const):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
(WebCore::RenderStyle::NonInheritedFlags::dumpDifferences const):
(WebCore::RenderStyle::InheritedFlags::dumpDifferences const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTextDecorationLine):
(WebCore::RenderStyle::initialTextDecorationLineInEffect):
(WebCore::RenderStyle::textDecorationLine const):
(WebCore::RenderStyle::textDecorationLineInEffect const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::addToTextDecorationLineInEffect):
(WebCore::RenderStyle::setTextDecorationLine):
(WebCore::RenderStyle::setTextDecorationLineInEffect):
* Source/WebCore/rendering/style/StyleInheritedData.cpp:
(WebCore::StyleInheritedData::StyleInheritedData):
(WebCore::StyleInheritedData::nonFastPathInheritedEqual const):
(WebCore::StyleInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleInheritedData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
(WebCore::StyleMiscNonInheritedData::StyleMiscNonInheritedData):
(WebCore::StyleMiscNonInheritedData::operator== const):
(WebCore::StyleMiscNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/SVGTextBoxPainter.cpp:
(WebCore::SVGTextBoxPainter&lt;TextBoxPath&gt;::paint):
(WebCore::positionOffsetForDecoration):
(WebCore::thicknessForDecoration):
(WebCore::SVGTextBoxPainter&lt;TextBoxPath&gt;::paintDecoration):
(WebCore::SVGTextBoxPainter&lt;TextBoxPath&gt;::paintDecorationWithStyle):
* Source/WebCore/rendering/svg/SVGTextBoxPainter.h:
* Source/WebCore/style/InlineTextBoxStyle.cpp:
(WebCore::minLogicalTopForTextDecorationLineUnder):
(WebCore::maxLogicalBottomForTextDecorationLineUnder):
(WebCore::enclosingRendererWithTextDecoration):
(WebCore::computedInkOverflowForDecorations):
(WebCore::inkOverflowForDecorations):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextDecorationLine): Deleted.
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::forwardInheritedValue):
* Source/WebCore/style/StyleExtractorConverter.h:
(WebCore::Style::ExtractorConverter::convertTextTransform):
(WebCore::Style::ExtractorConverter::convertTextDecorationLine): Deleted.
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::ExtractorCustom::extractTextDecorationShorthand):
(WebCore::Style::ExtractorCustom::extractTextDecorationShorthandSerialization):
* Source/WebCore/style/StyleExtractorSerializer.h:
(WebCore::Style::ExtractorSerializer::serializeTextDecorationLine): Deleted.
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.cpp: Added.
(WebCore::Style::TextDecorationLine::addOrReplaceIfNotNone):
(WebCore::Style::CSSValueConversion&lt;TextDecorationLine&gt;::operator):
(WebCore::Style::CSSValueCreation&lt;OptionSet&lt;TextDecorationLineFlags&gt;&gt;::operator):
(WebCore::Style::Serialize&lt;OptionSet&lt;TextDecorationLineFlags&gt;&gt;::operator):
(WebCore::Style::operator&lt;&lt;):
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h: Added.
(WebCore::Style::TextDecorationLine::TextDecorationLine):
(WebCore::Style::TextDecorationLine::isNone const):
(WebCore::Style::TextDecorationLine::isSpellingError const):
(WebCore::Style::TextDecorationLine::isGrammarError const):
(WebCore::Style::TextDecorationLine::isFlags const):
(WebCore::Style::TextDecorationLine::hasUnderline const):
(WebCore::Style::TextDecorationLine::hasOverline const):
(WebCore::Style::TextDecorationLine::hasLineThrough const):
(WebCore::Style::TextDecorationLine::hasBlink const):
(WebCore::Style::TextDecorationLine::containsAny const):
(WebCore::Style::TextDecorationLine::contains const):
(WebCore::Style::TextDecorationLine::remove):
(WebCore::Style::TextDecorationLine::switchOn const):
(WebCore::Style::TextDecorationLine::operator bool const):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView updateTextTouchBar]):

Canonical link: <a href="https://commits.webkit.org/299142@main">https://commits.webkit.org/299142@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/813d2e78225f9ea491ebfa8e205f0e1f3d9e24fd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117735 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37413 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28043 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123869 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69905 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/25b070a2-ab1b-4c98-8d21-2e827f408444) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119613 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38105 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45995 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89356 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59064 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6d0f5144-2ffa-4856-9afc-75624a993d93) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120687 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30330 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105567 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69850 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3ba77f24-0cf7-45b6-a061-92a3bbf70ef7) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29394 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23682 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67528 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99749 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23862 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126962 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44638 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33597 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98012 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44996 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101792 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97800 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24927 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43180 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21147 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41003 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44510 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50184 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43968 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47315 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45659 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/TextDecorationPainter.cpp:
```diff
@@ -35,6 +35,8 @@
 #include "TextBoxPainter.h"
 #include "TextRun.h"
+#include "rendering/style/RenderStyleConstants.h"
+#include "style/values/text-decoration/StyleTextDecorationLine.h"
```
```suggestion
#include "RenderStyleConstants.h"
#include "StyleTextDecorationLine.h"
```

nt1m commented on Source/WebCore/style/InlineTextBoxStyle.cpp:
```diff
@@ -279,23 +279,23 @@ GlyphOverflow inkOverflowForDecorations(const InlineIterator::LineBoxIterator& l
     }
 
-    auto underlineOffset = style.textDecorationLineInEffect().contains(TextDecorationLine::Underline)
+    auto underlineOffset = style.textDecorationLineInEffect().contains(TextDecorationLineFlags::Underline)
```
For here and the rest of the PR, if this is the same as `.hasUnderline()` we should use the shorter one if possible.

nt1m commented on Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.cpp:
```diff
@@ -0,0 +1,189 @@
+#include "css/CSSValue.h"
+#include "css/values/CSSValueTypes.h"
+#include "rendering/style/RenderStyle.h"
+#include "style/values/StyleValueTypes.h"
```
```suggestion
#include "CSSValue.h"
#include "CSSValueTypes.h"
#include "RenderStyle.h"
#include "StyleValueTypes.h"
```

nt1m commented on Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.cpp:
```diff
@@ -0,0 +1,189 @@
+
+namespace WebCore {
+namespace Style {
+void TextDecorationLine::addOrReplaceIfNotNone(const TextDecorationLine& value)
```
nit: add new line before

nt1m commented on Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.cpp:
```diff
@@ -0,0 +1,189 @@
+            return;
+        }
+    }
+    m_value = value.m_value;
```
Would be nice if you could use WTF::switchOn

nt1m commented on Source/WebCore/rendering/style/RenderStyle.cpp:
Using rare data seems like a bad idea for both of the text decoration line props, literally every link uses text-decoration-line

nt1m commented on Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
```diff
@@ -37,6 +37,8 @@
 #include "StyleTransformData.h"
 #include "StyleVisitedLinkColorData.h"
+#include "rendering/style/RenderStyle.h"
+#include "style/values/text-decoration/StyleTextDecorationLine.h"
```
```suggestion
#include "RenderStyle.h"
#include "StyleTextDecorationLine.h"
```

nt1m commented on Source/WebCore/rendering/style/StyleRareInheritedData.h:
```diff
@@ -230,6 +230,7 @@ class StyleRareInheritedData : public RefCounted<StyleRareInheritedData> {
     Style::ListStyleType listStyleType;
     Style::BlockEllipsis blockEllipsis;
 
+
```
nit: undo this change

vitorroriz committed:
text-decoration-line: Refactor TextDecorationLine into strong type
https://bugs.webkit.org/show_bug.cgi?id=297756
rdar://158897260

Reviewed by Tim Nguyen and Simon Fraser.

text-decoration-line used to be represented on Style as a OptionSet<TextDecorationLine>
and TextDecorationLine used to be an enum class.

We are refactoring TextDecorationLine into its own type aligned with other strong type changes.
The new type is now defined in StyleTextDecorationLine and it enforces the grammar:

text-decoration-line = none | [ underline || overline || line-through || blink ] | spelling-error | grammar-error

The new underlying representation for it is:

Variant<CSS::Keyword::None, CSS::Keyword::SpellingError, CSS::Keyword::GrammarError, OptionSet<TextDecorationLineFlags>> m_value;

This means that TextDecorationLine flags represents now only [ underline || overline || line-through || blink ]

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::LineDecorationStyle::LineDecorationStyle):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::onStyleChange):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::hasPlainText const):
(WebCore::AccessibilityRenderObject::hasUnderline const):
* Source/WebCore/accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::textAttributes const):
* Source/WebCore/accessibility/ios/AccessibilityObjectIOS.mm:
(WebCore::attributeStringSetStyle):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::fontAttributesAtSelectionStart):
* Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
(WebCore::updateAttributes):
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
(WebCore::Layout::logicalBottomForTextDecorationContent):
(WebCore::Layout::InlineDisplayContentBuilder::collectInkOverflowForTextDecorations):
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForHighlights):
* Source/WebCore/rendering/StyledMarkedText.cpp:
(WebCore::computeStyleForPseudoElementStyle):
(WebCore::computeStylesForTextDecorations):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter::paintForegroundAndDecorations):
(WebCore::computedTextDecorationType):
(WebCore::isDecoratingBoxForBackground):
(WebCore::TextBoxPainter::paintBackgroundDecorations):
(WebCore::TextBoxPainter::paintForegroundDecorations):
(WebCore::TextBoxPainter::paintPlatformDocumentMarkers):
* Source/WebCore/rendering/TextDecorationPainter.cpp:
(WebCore::TextDecorationPainter::paintBackgroundDecorations):
(WebCore::collectStylesForRenderer):
(WebCore::TextDecorationPainter::stylesForRenderer):
(WebCore::TextDecorationPainter::textDecorationsInEffectForStyle):
* Source/WebCore/rendering/TextDecorationPainter.h:
(WebCore::TextDecorationPainter::stylesForRenderer):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
(WebCore::RenderStyle::NonInheritedFlags::copyNonInheritedFrom):
(WebCore::RenderStyle::changeAffectsVisualOverflow const):
(WebCore::miscDataChangeRequiresRepaint):
(WebCore::RenderStyle::changeRequiresRepaint const):
(WebCore::RenderStyle::changeRequiresRepaintIfText const):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
(WebCore::RenderStyle::NonInheritedFlags::dumpDifferences const):
(WebCore::RenderStyle::InheritedFlags::dumpDifferences const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialTextDecorationLine):
(WebCore::RenderStyle::initialTextDecorationLineInEffect):
(WebCore::RenderStyle::textDecorationLine const):
(WebCore::RenderStyle::textDecorationLineInEffect const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::addToTextDecorationLineInEffect):
(WebCore::RenderStyle::setTextDecorationLine):
(WebCore::RenderStyle::setTextDecorationLineInEffect):
* Source/WebCore/rendering/style/StyleInheritedData.cpp:
(WebCore::StyleInheritedData::StyleInheritedData):
(WebCore::StyleInheritedData::nonFastPathInheritedEqual const):
(WebCore::StyleInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleInheritedData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
(WebCore::StyleMiscNonInheritedData::StyleMiscNonInheritedData):
(WebCore::StyleMiscNonInheritedData::operator== const):
(WebCore::StyleMiscNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/SVGTextBoxPainter.cpp:
(WebCore::SVGTextBoxPainter<TextBoxPath>::paint):
(WebCore::positionOffsetForDecoration):
(WebCore::thicknessForDecoration):
(WebCore::SVGTextBoxPainter<TextBoxPath>::paintDecoration):
(WebCore::SVGTextBoxPainter<TextBoxPath>::paintDecorationWithStyle):
* Source/WebCore/rendering/svg/SVGTextBoxPainter.h:
* Source/WebCore/style/InlineTextBoxStyle.cpp:
(WebCore::minLogicalTopForTextDecorationLineUnder):
(WebCore::maxLogicalBottomForTextDecorationLineUnder):
(WebCore::enclosingRendererWithTextDecoration):
(WebCore::computedInkOverflowForDecorations):
(WebCore::inkOverflowForDecorations):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertTextDecorationLine): Deleted.
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::forwardInheritedValue):
* Source/WebCore/style/StyleExtractorConverter.h:
(WebCore::Style::ExtractorConverter::convertTextTransform):
(WebCore::Style::ExtractorConverter::convertTextDecorationLine): Deleted.
* Source/WebCore/style/StyleExtractorCustom.h:
(WebCore::Style::ExtractorCustom::extractTextDecorationShorthand):
(WebCore::Style::ExtractorCustom::extractTextDecorationShorthandSerialization):
* Source/WebCore/style/StyleExtractorSerializer.h:
(WebCore::Style::ExtractorSerializer::serializeTextDecorationLine): Deleted.
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.cpp: Added.
(WebCore::Style::TextDecorationLine::addOrReplaceIfNotNone):
(WebCore::Style::CSSValueConversion<TextDecorationLine>::operator):
(WebCore::Style::CSSValueCreation<OptionSet<TextDecorationLineFlags>>::operator):
(WebCore::Style::Serialize<OptionSet<TextDecorationLineFlags>>::operator):
(WebCore::Style::operator<<):
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h: Added.
(WebCore::Style::TextDecorationLine::TextDecorationLine):
(WebCore::Style::TextDecorationLine::isNone const):
(WebCore::Style::TextDecorationLine::isSpellingError const):
(WebCore::Style::TextDecorationLine::isGrammarError const):
(WebCore::Style::TextDecorationLine::isFlags const):
(WebCore::Style::TextDecorationLine::hasUnderline const):
(WebCore::Style::TextDecorationLine::hasOverline const):
(WebCore::Style::TextDecorationLine::hasLineThrough const):
(WebCore::Style::TextDecorationLine::hasBlink const):
(WebCore::Style::TextDecorationLine::containsAny const):
(WebCore::Style::TextDecorationLine::contains const):
(WebCore::Style::TextDecorationLine::remove):
(WebCore::Style::TextDecorationLine::switchOn const):
(WebCore::Style::TextDecorationLine::operator bool const):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView updateTextTouchBar]):

Canonical link: https://commits.webkit.org/299142@main
